### PR TITLE
feat: add `AzureCosmosDBNoSqlVectorStore` and `SimpleCosmosDBReader`

### PR DESCRIPTION
**Pull request by amanrao23**
This PR makes following changes to the project:
- Add support for using Azure Cosmos DB NoSQL API as a vector-store.
- Add a simple Cosmos DB reader to read the data from the Cosmos DB.
- Add a shortStories JSON dataset.
- Add CosmosDBNoSqlVectorStore RAG example.

### CODE REVIEW

File: packages/llamaindex/tests/readers/SimpleCosmosDBReader.test.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import { CosmosClient } from "@azure/cosmos";
+import { Document } from "@llamaindex/core/schema";
+import {
+  SimpleCosmosDBReader,
+  type SimpleCosmosReaderLoaderConfig,
+} from "llamaindex";
+import { describe, expect, it } from "vitest";
+import { createMockClient } from "../utility/mockCosmosClient.js"; // Import the mock client
+
+describe("SimpleCosmosDBReader", () => {
```
Comment (manekinekko) [inline packages/llamaindex/tests/readers/SimpleCosmosDBReader.test.ts:10]: I noted some missing test cases like:
- wrong or invalid properties
- wrong or invalid values (null, undefined, booleans, numbers

File: packages/llamaindex/src/readers/SimpleCosmosDBReader.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  /**
+   * Flattens an array of strings or string arrays into a single-dimensional array of strings.
+   * @param texts - The array of strings or string arrays to flatten.
+   * @returns The flattened array of strings.
+   */
+  private flatten(texts: Array<string | string[]>): string[] {
+    return texts.flat();
+  }
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/src/readers/SimpleCosmosDBReader.ts]: Updated

File: packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+      getEnv("AZURE_COSMOSDB_NOSQL_CONNECTION_STRING");
+
+    const endpoint =
+      dbConfig.endpoint ?? getEnv("AZURE_COSMOSDB_NOSQL_ENDPOINT");
+
+    if (!dbConfig.client && !connectionString && !endpoint) {
+      throw new Error(
+        "AzureCosmosDBNoSQLVectorStore client, connection string or endpoint must be set.",
+      );
+    }
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts]: User has the option to pass one of cosmosClient/connectionString/endPoint. If none of these is configured, error is thrown in that case. I have updated the error message to be more informative now.

File: packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+      const node = metadataDictToNode(item["metadata"], {
+        fallback: {
+          id_: item["id"],
+          text: item["text"],
+          ...item["metadata"],
+        },
+      });
+      node.setContent(item["text"]);
+      const node_id = item["id"];
+      const node_score = item["SimilarityScore"];
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts]: Updated the code.

File: packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+      nodes.push(node);
+      ids.push(node_id);
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts]: Updated the code.

File: packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+      similarities.push(node_score);
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts]: Updated

File: packages/llamaindex/tests/vector-stores/AzureCosmosDBNoSqlVectorStore.int.test.ts [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    apiVersion: process.env.AZURE_OPENAI_EMBEDDING_API_VERSION,
+    endpoint: process.env.AZURE_OPENAI_EMBEDDING_ENDPOINT,
+    apiKey: process.env.AZURE_OPENAI_EMBEDDING_API_KEY,
+  },
+};
+
+Settings.llm = new OpenAI(llmInit);
+Settings.embedModel = new OpenAIEmbedding(embedModelInit);
+
+describe.skip("AzureCosmosDBNoSQLVectorStore", () => {
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/tests/vector-stores/AzureCosmosDBNoSqlVectorStore.int.test.ts:66]: This is disabled as this is an end to end test and required a CosmosDB account and Azure OpenAI credentials.

File: packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+  if (!endpoint || !accountKey) {
+    throw new Error(
+      "Invalid connection string: missing AccountEndpoint or AccountKey.",
+    );
+  }
+
+  return { endpoint, key: accountKey };
+}
+
+export class AzureCosmosDBNoSqlVectorStore
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts]: There might be addition of Azure CosmosDB Mongo API. To distinguish between these two, `NoSql` is required in the name.

File: packages/llamaindex/tests/readers/SimpleCosmosDBReader.test.ts (hunk 1 above)
Comment (amanrao23) [reply packages/llamaindex/tests/readers/SimpleCosmosDBReader.test.ts]: I have added a few more test cases

**Comment by amanrao23** [comment]
@himself65 please review this PR.

File: packages/llamaindex/src/readers/SimpleCosmosDBReader.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+  /**
+   * Loads data from a Cosmos DB container
+   * @returns {Promise<Document[]>}
+   */
+  public async loadData(
+    config: SimpleCosmosReaderLoaderConfig,
+  ): Promise<Document[]> {
+    const database = this.client.database(config.databaseName);
+    const container = database.container(config.containerName);
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/src/readers/SimpleCosmosDBReader.ts]: Added these changes.

File: packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+    for (const item of queryResults.resources) {
+      const node = metadataDictToNode(item["metadata"], {
+        fallback: {
+          id_: item["id"],
+          text: item["text"],
+          ...item["metadata"],
+        },
+      });
+      node.setContent(item["text"]);
+      const node_id = item["id"];
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts]: Updated

File: packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+  /**
+   * Initialize the CosmosDB container.
+   */
+  private async init(
+    client: CosmosClient,
+    databaseName: string,
+    containerName: string,
+    initOptions: AzureCosmosDBNoSQLInitOptions,
+  ): Promise<void> {
+    const { database } = await client.databases.createIfNotExists({
```
[the remark this answers is not in the document]
Comment (amanrao23) [reply packages/llamaindex/src/vector-store/AzureCosmosDBNoSqlVectorStore.ts:345]: This requirement will be mentioned in the docs.

**Review by himself65 — APPROVED** [review]

**Pull request merged**